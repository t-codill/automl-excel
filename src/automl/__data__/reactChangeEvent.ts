// tslint:disable: no-any
export const reactChangeEvent: React.ChangeEvent<any> = {
    target: {}
} as unknown as React.ChangeEvent<any>;
