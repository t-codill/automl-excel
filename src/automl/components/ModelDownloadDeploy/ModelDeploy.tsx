import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { CompoundButton } from "office-ui-fabric-react";
import * as React from "react";
import { PageNames } from "../../common/PageNames";
import { ModelDeployPanel } from "./ModelDeployPanel";

export interface IModelDeployProps {
    pageName: PageNames;
    experimentName: string | undefined;
    run: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined;
    parentRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDetailsDto | undefined;
    modelUri: string | undefined;
    modelId: string | undefined;
    modelName: string | undefined;
    onModelRegister(): void;
}

export interface IModelDeployState {
    showDeployPanel: boolean;
}

export class ModelDeploy extends React.Component<IModelDeployProps, IModelDeployState, {}> {
    constructor(props: IModelDeployProps) {
        super(props);
        this.state = {
            showDeployPanel: false
        };
    }
    public readonly render = (): React.ReactNode => {
        const style: React.CSSProperties = {
            boxSizing: "border-box",
            height: "80px",
            width: "48%",
        };
        return <>
            <CompoundButton
                style={style}
                primary={true}
                secondaryText={this.props.modelName}
                disabled={!this.props.modelUri}
                onClick={this.onModelDeployClick}>
                {this.props.pageName === PageNames.ParentRun ? "Deploy Best Model" : "Deploy Model"}
            </CompoundButton>
            {this.state.showDeployPanel &&
                <ModelDeployPanel
                    {...this.props}
                    onCancel={this.onModelDeployCancel}
                    onModelRegister={this.props.onModelRegister}
                />
            }
        </>;
    }

    private readonly onModelDeployClick = () => {
        this.setState({ showDeployPanel: true });
    }
    private readonly onModelDeployCancel = () => {
        this.setState({ showDeployPanel: false });
    }
}
