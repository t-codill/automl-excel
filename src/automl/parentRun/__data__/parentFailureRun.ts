import { NotNullableProperties } from "../../common/utils/NotNullableProperties";
import { IParentRunData } from "../ParentRun";

export const parentFailureRun: NotNullableProperties<IParentRunData> = {
    run: {
        runId: "AutoML_000",
        parentRunId: undefined,
        status: "Completed",
        startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
        endTimeUtc: new Date("2019-02-01T08:30:00.000Z"),
        properties: {
            AMLSettingsJsonString: "",
            primary_metric: "pm",
            errors: "Setup iteration failed: \\n \t\t\t\t\texception: \\n \t\t\t\t\t\tUnforseenConsequence('Invalid subject in the test chamber') \\n"
        }
    },
    experimentName: "Test",
    childRuns: [],
    childRunMetrics: [
        {
            spearman_correlation: 0.877770827248834
        }
    ]
};
