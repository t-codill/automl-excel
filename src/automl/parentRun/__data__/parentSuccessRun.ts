import { NotNullableProperties } from "../../common/utils/NotNullableProperties";
import { IParentRunData } from "../ParentRun";

export const parentSuccessRunAMLSettings = {
  primary_metric: "AUC_weighted",
  data_script: null,
  compute_target: "local",
  n_cross_validations: 3,
  max_cores_per_iteration: 1,
  blacklist_algos: [],
  exit_score: 100,
  name: "my_experiment",
  task_type: "classification"
};
export const parentSuccessRun: NotNullableProperties<IParentRunData> = {
  run: {
    runId: "AutoML_000",
    parentRunId: undefined,
    status: "Completed",
    startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
    endTimeUtc: new Date("2019-02-01T08:30:00.000Z"),
    properties: {
      AMLSettingsJsonString: JSON.stringify(parentSuccessRunAMLSettings)
    },
    tags: {
      iterations: "2"
    },
  },
  experimentName: "Test",
  childRuns: [
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_001",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      experimentId: "exp",
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_001",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_algorithm: "RandomForest",
        run_preprocessor: "MaxAbsScaler",
        score: "0.977770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_002",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_002",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_preprocessor: "MaxAbsScaler",
        score: "0.877770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_003",
      parentRunId: "AutoML_000",
      status: "Canceled",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_003",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        score: "0.877770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      }
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_004",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_004",
      properties: {
        runTemplate: "automl_child",
        pipeline_id: "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        run_template: "automl_child",
        run_algorithm: "MaxAbsScaler",
        score: "0.827770827248834",
        class_labels: "",
        primary_metric: "spearman_correlation",
      },
    },
    {
      createdUtc: new Date("2019-02-01T08:20:00.000Z"),
      runId: "AutoML_005",
      parentRunId: "AutoML_000",
      status: "Completed",
      startTimeUtc: new Date("2019-02-01T08:20:00.000Z"),
      endTimeUtc: new Date("2019-02-01T08:25:00.000Z"),
      options: {
        generateDataContainerIdIfNotSpecified: true
      },
      name: "AutoML_005",
    }
  ],
  childRunMetrics: [
    {
      spearman_correlation: 0.877770827248834
    }
  ]
};
