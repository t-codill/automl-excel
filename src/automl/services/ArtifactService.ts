import { ArtifactAPI, ArtifactAPIModels } from "@vienna/artifact";
import { v4 as uuid4 } from "uuid";
import { blob2string } from "../common/utils/blob2string";
import { getServiceCommonUrl } from "../common/utils/getServiceCommonUrl";
import { isRunCompleted } from "../common/utils/run";
import { ArtifactServicePath } from "./constants/ArtifactServicePath";
import { IServiceBaseProps } from "./ServiceBase";
import { ServiceBaseNonArm } from "./ServiceBaseNonArm";

const artifactClientApiVersion = "1.0";
const runOrigin = "ExperimentRun";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export interface IRunDataForArtifact {
    runId: string | undefined;
    parentRunId: string | undefined;
    status: string | undefined;
}

async function getArtifactContent(artifact: ArtifactAPIModels.GetArtifactContentByIdResponse): Promise<string | undefined> {
    const blobBody = await artifact.blobBody;
    if (blobBody) {
        return blob2string(blobBody);
    }
    return undefined;
}

function getContentUri(res: ArtifactAPIModels.ArtifactContentInformationDto | null): string | undefined {
    return res ? res.contentUri : undefined;
}

export class ArtifactService extends ServiceBaseNonArm<ArtifactAPI> {

    constructor(props: IServiceBaseProps) {
        super(props, ArtifactAPI, props.discoverUrls.history);
    }

    public async tryGetContentForRun(runId: string, path: string): Promise<string | null | undefined> {
        return this.tryGetContent({ origin: runOrigin, container: runId, path });
    }

    public async tryGetContent(artifact: IArtifact): Promise<string | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getContent(artifact: IArtifact): Promise<string | undefined> {
        return this.send(async (client, abortSignal) => {
            const artifactContent = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    path: artifact.path,
                    abortSignal
                });
            return getArtifactContent(artifactContent);
        });
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return this.parallelGetAllValues(params, async (param, client, abortSignal) => {
            const artifact = await client.getArtifactContentById2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                param.origin,
                param.container,
                artifactClientApiVersion, {
                    path: param.path,
                    abortSignal
                });
            return getArtifactContent(artifact);
        });
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return this.parallelGetAllValuesWithContinuationToken(runs, async (runId, client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async getAllArtifacts(runId: string): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllValuesWithContinuationToken(async (client, abortSignal, continuationToken) => {
            return client.getArtifactsInContainerOrPath2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                runOrigin,
                runId,
                artifactClientApiVersion, {
                    abortSignal,
                    continuationToken
                });
        });
    }

    public async tryGetArtifactContentInfo(artifact: IArtifact): Promise<ArtifactAPIModels.GetArtifactContentInformation2Response | null | undefined> {
        return this.trySend(async (client, abortSignal) => {
            return client.getArtifactContentInformation2(
                this.props.subscriptionId,
                this.props.resourceGroupName,
                this.props.workspaceName,
                artifact.origin,
                artifact.container,
                artifactClientApiVersion, {
                    abortSignal,
                    path: artifact.path
                });
        });
    }

    public async getModelUri(run: IRunDataForArtifact): Promise<string | null | undefined> {
        if (!run || !run.runId || !run.parentRunId || !isRunCompleted(run)) {
            return null;
        }
        const container = run.runId;
        const result = await this.trySend(
            async (client, abortSignal) => {
                return client.getArtifactContentInformation2(
                    this.props.subscriptionId,
                    this.props.resourceGroupName,
                    this.props.workspaceName,
                    runOrigin,
                    container,
                    artifactClientApiVersion, {
                        abortSignal,
                        path: ArtifactServicePath.Model
                    });
            });
        if (result === undefined) {
            return undefined;
        }
        if (!result || !result.contentUri) {
            return null;
        }
        return result.contentUri;
    }

    public async getDeployUri(run: IRunDataForArtifact): Promise<{
        modelUri?: string;
        condaUri?: string;
        scoringUri?: string;
    } | undefined> {
        if (!run || !run.runId || !run.parentRunId || !isRunCompleted(run)) {
            return {};
        }
        const container = run.runId;
        const result = await this.parallelTryGetAllValues(
            [
                ArtifactServicePath.Model,
                ArtifactServicePath.Conda,
                ArtifactServicePath.Scoring
            ],
            async (path, client, abortSignal) => {
                return client.getArtifactContentInformation2(
                    this.props.subscriptionId,
                    this.props.resourceGroupName,
                    this.props.workspaceName,
                    runOrigin,
                    container,
                    artifactClientApiVersion, {
                        abortSignal,
                        path
                    });
            });
        if (!result) {
            return undefined;
        }
        return {
            modelUri: getContentUri(result[0]),
            condaUri: getContentUri(result[1]),
            scoringUri: getContentUri(result[2])
        };
    }

    public async uploadArtifact(container: string, path: string, content: string): Promise<string | undefined> {
        return this.send(async (_client, abortSignal) => {
            const baseUrl = this.geBaseUrl();
            const artifactId = `${runOrigin}/${container}/${path}`;
            const url = `${baseUrl}/artifacts/content/${artifactId}?allowOverwrite=true`;
            const response = await window.fetch(new Request(url, {
                method: "POST",
                signal: abortSignal,
                body: new Blob([content]),
                headers: {
                    authorization: `Bearer ${this.props.getToken()}`,
                    "content-type": "application/octet-stream",
                    "x-ms-client-user-type": "AutoML WebUser",
                    "x-ms-client-request-id": uuid4(),
                    "x-ms-client-session-id": this.props.logger.getSessionId()
                }
            }));
            if (response.status === 200) {
                return artifactId;
            }
            return undefined;
        });
    }

    private geBaseUrl(): string {
        const prefix = `${this.props.discoverUrls.api}/artifact/v2.0`;
        const commonUrl = getServiceCommonUrl(this.props.subscriptionId, this.props.resourceGroupName, this.props.workspaceName);
        return `${prefix}${commonUrl}`;
    }
}
