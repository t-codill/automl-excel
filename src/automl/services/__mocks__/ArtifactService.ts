import { ArtifactAPIModels } from "@vienna/artifact";

export interface IArtifact {
    origin: string;
    container: string;
    path: string;
}

export class ArtifactService {
    private readonly tryGetContentResult = "tryGetContent";
    private readonly getAllArtifactsResult: ArtifactAPIModels.ArtifactDto[] = [
        {
            artifactId: "artifactId1",
            container: "container1",
            origin: "origin1",
            path: "path1"
        },
        {
            artifactId: "artifactId2",
            container: "container2",
            origin: "origin2",
            path: "path2"
        }
    ];
    private readonly tryGetArtifactUrlResult: ArtifactAPIModels.ArtifactContentInformationDto = {
        contentUri: "contentUri",
        path: "path"
    };
    private readonly getModelUrlsResult = ["getModelUrls1", "getModelUrls2"];

    public async tryGetContentForRun(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async tryGetContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getContent(): Promise<string | null | undefined> {
        return this.tryGetContentResult;
    }

    public async getAllContents(params: IArtifact[]): Promise<Array<string | undefined> | undefined> {
        return params.map((p) => `"getAllContentsResult:${p.origin}/${p.container}/${p.path}`);
    }

    public async getAllArtifacts(): Promise<ArtifactAPIModels.ArtifactDto[] | undefined> {
        return this.getAllArtifactsResult;
    }

    public async getAllArtifactsForRuns(runs: string[]): Promise<ArtifactAPIModels.ArtifactDto[][] | undefined> {
        return runs.map(() => this.getAllArtifactsResult);
    }

    public async tryGetArtifactUrl(): Promise<ArtifactAPIModels.ArtifactContentInformationDto | null | undefined> {
        return this.tryGetArtifactUrlResult;
    }

    public async getDeployUri(): Promise<{
        modelUri?: string;
        condaUri?: string;
        scoringUri?: string;
    } | undefined> {
        return {
            modelUri: "modelUri",
            condaUri: "condaUri",
            scoringUri: "scoringUri"
        };
    }

    public async getModelUri(): Promise<Array<string | null> | undefined> {
        return this.getModelUrlsResult;
    }

    public async uploadArtifact(_container: string, _path: string, _content: string): Promise<string | undefined> {
        return "mockArtifactId";
    }

    public dispose(): void {
        return;
    }

    public reset(): void {
        return;
    }
}
