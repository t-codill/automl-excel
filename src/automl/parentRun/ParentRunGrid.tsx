import * as React from "react";
import { shouldShowParentRunModel } from "../common/utils/run";
import { GridItem } from "../components/Grid/GridItem";
import { ChildRunIterationTable } from "./ChildRunIterationTable";
import { IParentRunData } from "./ParentRun";
import { ParentRunModel } from "./ParentRunModel";
import { ParentRunSkylineChart } from "./ParentRunSkylineChart";
import { ParentRunSummary } from "./ParentRunSummary";
import { ParentSettings } from "./ParentSettings";

export interface IParentRunGridProps extends IParentRunData {
    onModelDeploy(): void;
}
export class ParentRunGrid extends React.Component<IParentRunGridProps> {
    public readonly render = (): React.ReactNode => {
        const modelVisible = shouldShowParentRunModel(this.props.run && this.props.run.status);
        return (
            <div className="ms-Grid" dir="ltr">
                <div className="ms-Grid-col ms-sm12 ms-md12 ms-lg12 ms-xl9">
                    <GridItem title="Iteration Chart">
                        <ParentRunSkylineChart  {...this.props} />
                    </GridItem>
                    <GridItem title="Iterations">
                        <ChildRunIterationTable  {...this.props} />
                    </GridItem>
                </div>
                <div className="ms-Grid-col ms-sm12 ms-md12 ms-lg12 ms-xl3">
                    {modelVisible && <ParentRunModel {...this.props} />}
                    <GridItem title="Run Summary">
                        <ParentRunSummary run={this.props.run} />
                    </GridItem>
                    <GridItem title="Run Settings">
                        <ParentSettings run={this.props.run} />
                    </GridItem>
                </div>
            </div>
        );
    }
}
