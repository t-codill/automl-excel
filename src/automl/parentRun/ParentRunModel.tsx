import { RunHistoryAPIsModels } from "@vienna/runhistory";
import { isEqual, orderBy } from "lodash";
import * as React from "react";
import { PageNames } from "../common/PageNames";
import { isRunCompleted } from "../common/utils/run";
import { safeParseJson } from "../common/utils/safeParseJson";
import { ModelSection } from "../components/ModelDownloadDeploy/ModelSection";
import { PageLoadingSpinner } from "../components/Progress/PageLoadingSpinner";
import { IParentRunData } from "./ParentRun";
import { IParentRunGridProps } from "./ParentRunGrid";

export interface IParentRunModelState {
    bestRun: RunHistoryAPIsModels.MicrosoftMachineLearningRunHistoryContractsRunDto | undefined;
}

export class ParentRunModel extends React.PureComponent<IParentRunGridProps, IParentRunModelState> {
    constructor(props: IParentRunGridProps) {
        super(props);
        this.state = {
            bestRun: this.getBestRun()
        };
    }

    public async componentDidUpdate(prevProps: IParentRunData): Promise<void> {
        if (isEqual(this.props.childRuns, prevProps.childRuns)) {
            return;
        }
        this.setState({
            bestRun: this.getBestRun()
        });
    }

    public readonly render = (): React.ReactNode => {
        if (!this.state.bestRun || !this.props.run || !this.props.experimentName) {
            return <PageLoadingSpinner />;
        }

        return <ModelSection
            experimentName={this.props.experimentName}
            pageName={PageNames.ParentRun}
            parentRun={this.props.run}
            run={this.state.bestRun}
            onModelDeploy={this.props.onModelDeploy}
        />;
    }

    private readonly getBestRun = () => {
        if (!this.props.childRuns || !this.props.run || !this.props.run.properties) {
            return undefined;
        }
        const min = safeParseJson(this.props.run.properties.AMLSettingsJsonString).metric_operation === "minimize";
        const scoredRuns = this.props.childRuns.map((r) => ({
            ...r,
            runId: r.runId,
            parentRunId: r.parentRunId,
            status: r.status,
            score: r.properties && parseFloat(r.properties.score)
        }));

        const sortedRuns = orderBy(scoredRuns, ["score"], [min ? "asc" : "desc"]);
        for (const bestRun of sortedRuns) {
            if (bestRun.score === undefined || isNaN(bestRun.score) || !isRunCompleted(bestRun)) {
                continue;
            }
            return bestRun;
        }
        return undefined;
    }
}
