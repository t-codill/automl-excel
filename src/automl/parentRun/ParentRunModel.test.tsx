import { shallow, ShallowWrapper } from "enzyme";
import * as React from "react";
import { ModelDeploy } from "../components/ModelDownloadDeploy/ModelDeploy";
import { ArtifactService } from "../services/ArtifactService";
import { parentSuccessRun, parentSuccessRunAMLSettings } from "./__data__/parentSuccessRun";
import { IParentRunGridProps } from "./ParentRunGrid";
import { IParentRunModelState, ParentRunModel } from "./ParentRunModel";

jest.mock("../services/ArtifactService");

describe("Parent Run Model", () => {
    const onModelRegister = jest.fn();
    let tree: ShallowWrapper<IParentRunGridProps, IParentRunModelState>;
    beforeAll(async () => {
        tree = shallow(
            <ParentRunModel
                experimentName="foo"
                run={undefined}
                childRuns={undefined}
                childRunMetrics={undefined}
                onModelRegister={onModelRegister} />
        );
        await Promise.resolve();
    });

    it("should render pageLoadingSpinner with undefined", async () => {
        expect(tree)
            .toMatchInlineSnapshot("<pageLoadingSpinner />");
    });

    it("should render with success run", async () => {
        tree.setProps({
            run: parentSuccessRun.run,
            childRuns: parentSuccessRun.childRuns,
            childRunMetrics: parentSuccessRun.childRunMetrics
        });
        await Promise.resolve();
        expect(tree)
            .toMatchSnapshot();
    });

    it("should refresh when child run changes", async () => {
        tree.setProps({ childRuns: [] });
        await Promise.resolve();
        expect(tree.state("uri"))
            .toBeUndefined();
    });

    it("should existing model id if it is registered already", async () => {
        const childRuns = [
            {
                runNumber: 1,
                properties: {
                    score: "1"
                },
                tags: {
                    model_id: "model_id_registered"
                }
            }
        ];
        tree.setProps({ childRuns });
        await Promise.resolve();
        expect(tree.find(ModelDeploy)
            .prop("modelId"))
            .toBe("model_id_registered");
    });

    it("should pick max value", async () => {
        const childRuns = [
            {
                properties: {
                    score: "1"
                }
            },
            {
                properties: {
                    score: "2"
                }
            }
        ];
        tree.setProps({ childRuns });
        await Promise.resolve();
        expect(tree.state("bestRun"))
            .toEqual({
                parentRunId: undefined,
                properties: { score: "2" },
                runId: undefined,
                score: 2,
                status: undefined
            });
    });

    it("should pick min value", async () => {
        const childRuns = [
            {
                properties: {
                    score: "1"
                }
            },
            {
                properties: {
                    score: "2"
                }
            },
            {
                properties: {
                    score: "3"
                }
            }
        ];
        tree.setProps({
            childRuns,
            run: {
                ...parentSuccessRun.run,
                properties: {
                    ...parentSuccessRun.run.properties,
                    AMLSettingsJsonString: JSON.stringify({
                        ...parentSuccessRunAMLSettings,
                        primary_metric: "normalized_root_mean_squared_error",
                        metric_operation: "minimize"
                    })
                }
            }
        });
        await Promise.resolve();
        expect(tree.state("bestRun"))
            .toEqual({
                parentRunId: undefined,
                properties: { score: "1" },
                runId: undefined,
                score: 1,
                status: undefined
            });
    });

    it("should skip model without uri", async () => {
        const childRuns = [
            {
                properties: {
                    score: "1"
                }
            },
            {
                properties: {
                    score: "2"
                }
            },
            {
                properties: {
                    score: "3"
                }
            },
            {
                properties: {
                    score: "4"
                }
            }
        ];
        const getModelUrlSpy = jest.spyOn(ArtifactService.prototype, "getModelUrl");
        getModelUrlSpy.mockReturnValueOnce(Promise.resolve(null));
        tree.setProps({
            childRuns,
        });
        await Promise.resolve();
        await Promise.resolve();
        expect(tree.state("bestRun"))
            .toEqual({
                parentRunId: undefined,
                properties: { score: "2" },
                runId: undefined,
                score: 2,
                status: undefined
            });
    });
});
