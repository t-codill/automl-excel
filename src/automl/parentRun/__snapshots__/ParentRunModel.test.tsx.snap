// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parent Run Model should render with success run 1`] = `
<ModelSection
  experimentName="foo"
  onModelDeploy={[MockFunction]}
  pageName="ParentRun"
  parentRun={
    Object {
      "endTimeUtc": 2019-02-01T08:30:00.000Z,
      "parentRunId": undefined,
      "properties": Object {
        "AMLSettingsJsonString": "{\\"primary_metric\\":\\"AUC_weighted\\",\\"data_script\\":null,\\"compute_target\\":\\"local\\",\\"n_cross_validations\\":3,\\"max_cores_per_iteration\\":1,\\"blacklist_algos\\":[],\\"exit_score\\":100,\\"name\\":\\"my_experiment\\",\\"task_type\\":\\"classification\\"}",
      },
      "runId": "AutoML_000",
      "startTimeUtc": 2019-02-01T08:20:00.000Z,
      "status": "Completed",
      "tags": Object {
        "iterations": "2",
      },
    }
  }
  run={
    Object {
      "createdUtc": 2019-02-01T08:20:00.000Z,
      "endTimeUtc": 2019-02-01T08:25:00.000Z,
      "experimentId": "exp",
      "name": "AutoML_001",
      "options": Object {
        "generateDataContainerIdIfNotSpecified": true,
      },
      "parentRunId": "AutoML_000",
      "properties": Object {
        "class_labels": "",
        "pipeline_id": "eda663ea8f36e6c07948c99bba9d1becb40991a5",
        "primary_metric": "spearman_correlation",
        "runTemplate": "automl_child",
        "run_algorithm": "RandomForest",
        "run_preprocessor": "MaxAbsScaler",
        "run_template": "automl_child",
        "score": "0.977770827248834",
      },
      "runId": "AutoML_001",
      "score": 0.977770827248834,
      "startTimeUtc": 2019-02-01T08:20:00.000Z,
      "status": "Completed",
    }
  }
/>
`;
